### PR TITLE
Export in-browser editable HTML files that can be saved to disk

### PR DESCRIPTION
**Pull request by alejoluc**
This pull requests adds the feature of exporting editable HTML files. Specifically, the editable parts are the annotation cells, not the highlights. The exported HTML now includes a script that, when it notices a change, will display a message allowing the user to save or discard the changes.

The changes are saved by downloading a new html file that includes the newly edited content. I have looked around self-editing html projects, most notably TiddlyWiki, and this seems to be the way they are doing it.

If the user has set the browser to ask him where to download a file, then he can easily overwrite the old file.

Here's a screenshot of the message that pops-up whenever the user edits anything in the annotation cells.

screenshot

The newly generated file is, of course, also editable in the same way as the file exported by the Windows application.

### CODE REVIEW

File: ToxAnnotationsExporter/app.config [hunk 1]
```diff
@@ -52,23 +52,29 @@
...
+#saveLayer { display:none; position: fixed; top: 0; left: 0; width: 100%; margin: 0; padding: 10px; background-color: #7f93c4; color: #FFFFFF; text-align: center; }
+#saveLayer a { color: #FFFFFF; }
 table, tr, td, th { border-style: solid; border-width: 1px; border-color: #055f9d; border-collapse: collapse; }
 th { text-align: left; vertical-align: top; padding: 5px; }
 td { text-align: justify; vertical-align: top; padding: 5px; }
 p { text-align: justify;  margin: 0; text-indent: 1em; }
 span.AnnotationHighlight { background-color: yellow }</value>
             </setting>
+            <setting name="HtmlExport_SaveLayer" serializeAs="String">
+              <value>[div id="saveLayer"]The contents of the annotations have been modified, do you wish to [a id="saveLink" href="javascript:save()"]save the changes (download new file)[/a] or [a id="discardLink" href="javascript:discard()"]discard them (reload the page)[/a]?[/div]</value>
```
Comment (alejoluc) [inline ToxAnnotationsExporter/app.config:65]: Here's the "save dialog" message

**Comment by alejoluc** [comment]
I have just added a fix that makes this work correctly when exporting highlights for several books into separate files

**Comment by alejoluc** [comment]
I have just added another fix that changes nothing in the behavior but makes the code cleaner. This pull request is ready so you can merge if you think it's good enough. You can test it by cloning my fork of the repository and building the binary.

**Comment by Toxaris-NL** [comment]
I am in principle fine with the changes. I am doubting however if the settings are a good place to store the javascript. I am thinking that perhaps adding the script as a resource and pull it from there when needed is better and easier to maintain.

**Comment by alejoluc** [comment]
I have thought about that too and it was the initial approach, however, I decided in going for the settings file since as of now everything is stored there and the binary ships without additional resources (like the CSS). Admittedly, the JavaScript code is more complex than the CSS snippet, so it may guarantee an additional resource. I can change the code to use an external JavaScript file. It has the additional benefit that the user can tinker with it if he so wishes.

**Comment by alejoluc** [comment]
I have already done it but I'm having a hard time stopping git ignoring the script file, since it's going to be in one of the bin/ subfolders. I have tried many rules, from specifically stating which file I want to include to telling it to NOT ignore any .js files, but no luck

**Comment by alejoluc** [comment]
Done. I have edited the .gitignore file. Turns out when you ignore a whole folder, if you want to include a subfolder or file that is deep withing the hierarchy, you have to un-ignore every parent until you reach it, for every ignore rule you write. That's why I deleted some ignore rules, but I made sure they had no effect when adding the files and only the files inside Resources/ (ie the single script) were added

**Pull request merged**